neo/Network/LocalNode.cs

```diff
@@ -177,22 +184,9 @@ public static void AllowHashes(IEnumerable<UInt256> hashes)
 
         private void Blockchain_PersistCompleted(object sender, Block block)
         {
-            Transaction[] remain;
-            lock (mem_pool)
-            {
-                foreach (Transaction tx in block.Transactions)
-                {
-                    mem_pool.Remove(tx.Hash);
-                }
-                if (mem_pool.Count == 0) return;
+            foreach (Transaction tx in block.Transactions)
+                persistedTxHashQueue.Enqueue(tx.Hash);
```

Comment:
The code before was broken in that way anyway; there was a gap between block being persisted and the mempool being updated -- the only gaurantee before was that you would have the transactions from a whole number of blocks, but it didn't gaurantee anything about having transactions from the latest block. If you want the code to be consistent in that way the code as it is needs to change also. Let's chat. 
Can you send me a message on telegram to @jsolman, maybe we can get an audio call going. I'd like to discuss the matter with you further directly if possible.

Reply:
To be more clear: the problem you are indicating can happen with the current code before my change; nothing gaurantees with the current code that the persist thread that locks the mem_pool and updates the transactions will not be blocked for a while due to thread scheduling and that other methods could read what you are referring to as the "wrong data" (which by that I think you mean outdated data) such as in AddTransaction that you refrenced.

Reply:
If we want the transactions for a block in the mem_pool to come in atomically, I can add a second queue and we can process transactions that are from a block (using `newBlockQueue`) independently of transactions from AddTransaction, and only read from the `newBlockQueue` once all transactions from the block have been written to the queue; however, that doesn't really do anything to prevent transactions getting in the mem_pool that shouldn't be valid in the case that a new block comes in after the individual transactions have already been added; so it seems like doing that work is moot.

Reply:
Another route that will potentially improve things and make the change that I better than the previous code would be if I add another high priority thread that just updates the mem_pool right away from the queue I added; that will potentially stop invalid transactions sooner. In fact that is the right thing to do for sure; it will be better than the existing code today.

I'll add that change also and update the pull request.

Reply:
Actually, I'm still not seeing much of a point though since the mem_pool is locked by the AddTransactionLoop thread, the other thread I mentioned last wouldn't be able to acquire a lock to update it while transaction verification is taking place. I look forward to discussing further with you directly.

Reply:
As it is today, after transactions are persisted there is a window of time before the mem_pool is written, where an AddTransaction could be called and have an inconsistent view of the mem_pool since it will still contain the already persisted transactions. Don’t you agree? How is that problem any worse than with this change of having potentially half of them removed? I’m going to wait for your reply before I make any additional code change, since I want to understand what problem we are trying to solve.

Reply:
In the latest commit on this pull request I fixed the issue of the blockchain not being locked while transactions are being persisted and while they are being verified.